### ISSUE TITLE
minions appear to be ignoring the 'order' in highdata

### ISSUE THREAD

**Issue by gtmanfred** [saltstack/salt#26234]
It does not appear to be correctly using the order to run the states.

```
----------
          ID: redis30u
    Function: service.running
        Name: redis
      Result: True
     Comment: Service redis has been enabled, and is running
     Started: 05:18:55.545201
    Duration: 710.252 ms
     Changes:   
              ----------
              redis:
                  True
----------
          ID: redis slaveof
    Function: file.append
        Name: /etc/redis.conf
      Result: True
     Comment: Appended 1 lines
     Started: 05:18:56.255867
    Duration: 7.221 ms
     Changes:   
              ----------
              diff:
                  --- 
                  +++ 
                  @@ -936,3 +936,4 @@
                   # in order to commit the file to the disk more incrementally and avoid
                   # big latency spikes.
                   aof-rewrite-incremental-fsync yes
                  +slaveof 172.16.0.13 6379
----------
          ID: redis-sentinel
    Function: file.replace
        Name: /etc/redis-sentinel.conf
      Result: True
     Comment: Changes were made
     Started: 05:18:56.263374
    Duration: 7.2 ms
     Changes:   
              ----------
              diff:
                  --- 
                  +++ 
                  @@ -50,7 +50,7 @@
                   #
                   # Note: master name should not include special characters or spaces.
                   # The valid charset is A-z 0-9 and the three characters ".-_".
                  -sentinel monitor mymaster 127.0.0.1 6379 2
                  +sentinel monitor mymaster 172.16.0.13 6379 2
                   
                   # sentinel auth-pass <master-name> <password>
                   #
----------
          ID: redis-sentinel
    Function: service.running
      Result: True
     Comment: Service redis-sentinel has been enabled, and is running
     Started: 05:18:56.270827
    Duration: 687.584 ms
     Changes:   
              ----------
              redis-sentinel:
                  True
----------
          ID: nginx
    Function: pkg.installed
      Result: True
     Comment: The following packages were installed/updated: nginx
     Started: 05:18:56.959652
    Duration: 11075.971 ms
     Changes:   
              ----------
              GeoIP:
                  ----------
                  new:
                      1.5.0-9.el7
                  old:
              fontconfig:
                  ----------
                  new:
                      2.10.95-7.el7
                  old:
              fontpackages-filesystem:
                  ----------
                  new:
                      1.44-8.el7
                  old:
              gd:
                  ----------
                  new:
                      2.0.35-26.el7
                  old:
              gperftools-libs:
                  ----------
                  new:
                      2.1-1.el7
                  old:
              libXpm:
                  ----------
                  new:
                      3.5.10-5.1.el7
                  old:
              libunwind:
                  ----------
                  new:
                      1.1-10.el7
                  old:
              libxslt:
                  ----------
                  new:
                      1.1.28-5.el7
                  old:
              nginx:
                  ----------
                  new:
                      1.6.3-6.el7
                  old:
              nginx-filesystem:
                  ----------
                  new:
                      1.6.3-6.el7
                  old:
----------
          ID: nginx
    Function: service.running
      Result: True
     Comment: Service nginx has been enabled, and is running
     Started: 05:19:08.227116
    Duration: 834.341 ms
     Changes:   
              ----------
              nginx:
                  True
----------
          ID: nginx
    Function: user.present
      Result: False
     Comment: The following group(s) are not present: taplist
     Started: 05:19:09.062484
    Duration: 1.243 ms
     Changes:   
----------
          ID: listener_redis30u
    Function: service.mod_watch
        Name: redis
      Result: True
     Comment: Service restarted
     Started: 05:19:09.064325
    Duration: 271.343 ms
     Changes:   
              ----------
              redis:
                  True
----------
          ID: listener_redis-sentinel
    Function: service.mod_watch
        Name: redis-sentinel
      Result: True
     Comment: Service restarted
     Started: 05:19:09.336013
    Duration: 246.407 ms
     Changes:   
              ----------
              redis-sentinel:
                  True

Summary for local
-------------
Succeeded: 29 (changed=28)
Failed:     1
-------------
Total states run:     30
Total run time:   45.299 s
```

This is the output of where everything fails.

**Comment by gtmanfred** [comment]
This is the top file.

https://github.com/taplists/taplist-salt/blob/txlf/top.sls

So app should be run between redis and nginx.

from state.show_highstate

```
    basic user home:
        ----------
        __env__:
            base
        __sls__:
            app
        file:
            |_
              ----------
              name:
                  /home/taplist
            |_
              ----------
              mode:
                  755
            |_
              ----------
              user:
                  taplist
            |_
              ----------
              group:
                  taplist
            - directory
            |_
              ----------
              order:
                  10037
```

createing the user is order 10037 but the nginx user is 10025

```
        user:
            |_
              ----------
              name:
                  nginx
            |_
              ----------
              groups:
                  - taplist
            - present
            |_
              ----------
              order:
                  10025
```

Which is causing it to fail, even though app is higher up in top.sls than nginx, the nginx user change is being run before the user.present in app.

I have tried this on 2015.5 2015.8 and develop, and gotten the same results... I am not sure if I am missing something or not, cause this used to work.

**Comment by gtmanfred** [comment]
Now, here is the one that is really confusing.

```
----------
          ID: taplist config uwsgi
    Function: file.managed
        Name: /etc/uwsgi.d/taplists.beer.ini
      Result: False
     Comment: User uwsgi is not available Group uwsgi is not available
     Started: 05:45:03.818507
    Duration: 1.172 ms
     Changes:   
----------
```

Creating the config file fails, and it is put at the bottom of a file where the user.present for uwsgi should have already run.

https://github.com/taplists/taplist-salt/blob/txlf/uwsgi/init.sls

```
----------
          ID: taplist config nginx
    Function: file.managed
        Name: /etc/nginx/conf.d/taplists.beer.conf
      Result: True
     Comment: File /etc/nginx/conf.d/taplists.beer.conf updated
     Started: 05:53:01.458884
    Duration: 25.756 ms
     Changes:   
              ----------
              diff:
                  New file
              mode:
                  0644
----------
          ID: restart nginx
    Function: service.running
        Name: nginx
      Result: True
     Comment: The service nginx is already running
     Started: 05:53:01.484932
    Duration: 201.639 ms
     Changes:   
----------
          ID: restart sshd
    Function: service.running
        Name: sshd
      Result: True
     Comment: The service sshd is already running
     Started: 05:53:01.686925
    Duration: 201.294 ms
     Changes:   
----------
          ID: taplist config uwsgi
    Function: file.managed
        Name: /etc/uwsgi.d/taplists.beer.ini
      Result: False
     Comment: User uwsgi is not available Group uwsgi is not available
     Started: 05:53:01.891120
    Duration: 1.105 ms
     Changes:   
----------
          ID: listener_restart nginx
    Function: service.mod_watch
        Name: nginx
      Result: True
     Comment: Service restarted
     Started: 05:53:01.892734
    Duration: 273.055 ms
     Changes:   
              ----------
              nginx:
                  True

Summary for local
-------------
Succeeded: 36 (changed=4)
Failed:     1
-------------
Total states run:     37
Total run time:    4.622 s
```


I have also confirmed that this user.present state hasn't run, even though it has a lower order than the ini  file.

```
    taplist config uwsgi:
        ----------
        __env__:
            base
        __sls__:
            uwsgi
        file:
            |_
              ----------
              name:
                  /etc/uwsgi.d/taplists.beer.ini
            |_
              ----------
              source:
                  salt://uwsgi/files/taplists.beer.ini
            |_
              ----------
              user:
                  uwsgi
            |_
              ----------
              group:
                  uwsgi
            - managed
            |_
              ----------
              order:
                  10047
```

```
        user:
            |_
              ----------
              name:
                  uwsgi
            |_
              ----------
              groups:
                  - taplist
            - present
            |_
              ----------
              order:
                  10045
```

Which makes this super weird.

Thanks.
Daniel

**Comment by gtmanfred** [comment]
If someone needs access to my environment to confirm this, let me know.

**Comment by ryanwalder** [comment]
Are you using 2015.5.3?

I've noticed that requisites seem to be broken in 2015.5.3 for me: 
https://github.com/saltstack/salt/issues/26241

I've fixed by rolling back to 2015.5.2 till there is a fix.

**Comment by gtmanfred** [comment]
I am not using any requisites.

I am only using the top down file order, and then the fact that top.sls is ordered.

**Comment by gtmanfred** [comment]
2015.5.2 does in fact work apparently.

**Comment by jfindlay** [comment]
@gtmanfred, thanks for the report.

**Comment by gtmanfred** [comment]
In 2015.5.2, this is the ordering for the above states that were broken.

```
        user:
            |_
              ----------
              names:
                  - taplist
            - present
            |_
              ----------
              order:
                  10024
```
```
        user:
            |_
              ----------
              name:
                  nginx
            |_
              ----------
              groups:
                  - taplist
            - present
            |_
              ----------
              order:
                  10032
```
```
        pkg:
            |_
              ----------
              name:
                  uwsgi-plugin-python
            - installed
            |_
              ----------
              order:
                  10039
        user:
            |_
              ----------
              name:
                  uwsgi
            |_
              ----------
              groups:
                  - taplist
            - present
            |_
              ----------
              order:
                  10041
```
```
    taplist config uwsgi:
        ----------
        __env__:
            base
        __sls__:
            uwsgi
        file:
            |_
              ----------
              name:
                  /etc/uwsgi.d/taplists.beer.ini
            |_
              ----------
              source:
                  salt://uwsgi/files/taplists.beer.ini
            |_
              ----------
              user:
                  uwsgi
            |_
              ----------
              group:
                  uwsgi
            - managed
            |_
              ----------
              order:
                  10043
```


So everything is in the correct order and follows the order.

It looks like I grabbed the wrong state id in one of my previous updates, so I am going to check that one again, and make sure that the state order is in fact higher or not.

**Comment by gtmanfred** [comment]
looks like it is putting it in the wrong order still.

```
        user:
            |_
              ----------
              names:
                  - taplist
            - present
            |_
              ----------
              order:
                  10035
```
```
        user:
            |_
              ----------
              name:
                  nginx
            |_
              ----------
              groups:
                  - taplist
            - present
            |_
              ----------
              order:
                  10024
```

**Comment by gtmanfred** [comment]
@jfindlay no problem, if there is any way yall can get this cleaned up, I am trying to demo a bunch of the ordering stuff next week at the Texas Linux Fest, and would like to be able to use some of the things in 2015.8.

Thanks
Daniel

**Comment by ryan-lane** [comment]
This seems like a pretty major regression. Are there no tests for this? Can tests be added for this?

**Comment by jfindlay** [comment]
@gtmanfred, @ryan-lane, I'll see if I can work on this in the next day or two and also write up some tests.

**Comment by UtahDave** [comment]
@gtmanfred would it be difficult for your current project to use requisites as a workaround until this ordering issue gets fixed?

**Comment by cachedout** [comment]
@gtmanfred Is there any chance you can `git bisect` through develop to see if you can find where this was introduced?

**Comment by gtmanfred** [comment]
@UtahDave That is what I am planning on doing if I have to :+1:

@cachedout  I am working on a git bisect script so that I can just let it run on it's own, but I will probably not finish that up until tonight.  Can I just run it on 2015.5.2 to 2015.5.3, do you think that will be good enough to get you enough information for 2015.8 and develop?

**Comment by cachedout** [comment]
@gtmanfred Anything will help, of course. Obviously, an exact commit would be ideal but I'll take what I can get. :]

**Comment by gtmanfred** [comment]
Sounds good, I am going to start with 2015.5.2 -> 2015.5.3, and once I have that, I will run one for develop.

**Comment by cachedout** [comment]
@gtmanfred Awesome! That should get us on our way. :]

**Comment by gtmanfred** [comment]
@cachedout someone gave me a great Idea to test it it happens with masterless minions, and then run it all through docker containers, otherwise I would have to build new servers, so I am working on that now.

**Comment by cachedout** [comment]
@gtmanfred Ah, yes. That is the way I usually test these sorts of things myself. Sorry, I should have mentioned that. :]

**Comment by gtmanfred** [comment]
I was unable to replicate it in the docker enviroment, so I said eff it, and tried building a new master to see if it was reproducible, and it still happens.

I am going to work on a script in the morning to see if I can bisect it by building cloud servers.

**Comment by cachedout** [comment]
Darn, OK. For the sake of completness, I should point out that we did recently merge a top-file fix but I really can't see how it would affect this particular situation. Regardless, that is here if you want to try reverinting it locally: https://github.com/saltstack/salt/pull/24878 (I'd be surprised if this were it, but I've been surprised before.) ;]

**Comment by gtmanfred** [comment]
That doesn't appear to be in 2015.5.3, so I doubt it would affect this.

I have the script almost ready, I am about to launch it in a tmux session, and go to bed and see if it comes up with anything in the morning.

**Comment by cachedout** [comment]
Whew! (Since that was my PR). ;]

Thanks for your good investigative work here, @gtmanfred. Keep us posted and let's talk tomorrow.

**Comment by gtmanfred** [comment]
![I can't even](http://i.imgur.com/k5SyHiV.gif)

I am doing it manually now.

**Comment by gtmanfred** [comment]
```
[root@salt2 salt]# git bisect good
f0f512a4da0f14b319b36418b7360bd7a5953fdd is the first bad commit
[root@salt2 salt]# git show f0f512a4da0f14b319b36418b7360bd7a5953fdd
commit f0f512a4da0f14b319b36418b7360bd7a5953fdd
Merge: 21709aa 3187d5d
Author: Justin Findlay <jfindlay@gmail.com>
Date:   Wed Jul 1 13:45:56 2015 -0600

    Merge pull request #25109 from jfindlay/2015.5

    add 2015.5.3 release notes
```

![wat](http://i.imgur.com/CwtNm6L.png)

**Comment by cachedout** [comment]
Heh. Those darn release notes! ;]

**Comment by gtmanfred** [comment]
ಠ_ಠ and now it is working. :/

**Comment by jfindlay** [comment]
Who needs release notes anyway?

**Comment by gtmanfred** [comment]
It still fails in develop.

**Comment by gtmanfred** [comment]
I think it is an error that requires something on both the minion and the master, haveing master on develop, but the minion on 2015.5.3 isn't causing the error, but both on 2015.5.3 is.

But then both on develop causes the same problem in the same exact way.

**Comment by gtmanfred** [comment]
```
[root@salt salt]# git bisect bad
Bisecting: 55 revisions left to test after this (roughly 6 steps)
[5eda47aa8451ff8cc21d75cc9dd5e82421411701] Merge pull request #26145 from sjorge/imgadm-output-cleanup
```
First bad one, finally getting somewhere!

**Comment by gtmanfred** [comment]
@cachedout 
```
cfb698222713257f68a8697403bc67cea7b87fff is the first bad commit
commit cfb698222713257f68a8697403bc67cea7b87fff
Author: Mike Place <mp@saltstack.com>
Date:   Mon Jun 15 11:30:14 2015 -0600

    Adding some dictupdate merge tests

    Add new option for top_file_merging_strategy to config

    Add master config file option

    Adding additional tests

    Lexicographical test works

    Make default strategy

    Simplify to just an option about whether or not to merge

    Changes

    Implement ordered top files

    Implement merge strategies and default top files

    Lint

    Doc rewrite and start on ordering

    Remove debugging

    Moving to env_order

    Adding env_order

    Still writing tests

    More tests and docs

:040000 040000 d62201b30e217e862e4ab9226e099ab64567df73 5a2f00a1bb36fadc3186bf78b33b6751cdb43df4 M	conf
:040000 040000 0cf5fffe554674203a4190c3658298f2763bb84f cb5cb5b77488b7f99a6df2ba398927a29a27d918 M	doc
:040000 040000 2c7daf49872a0fe74989acb367cef5a6dd735264 2bb1490784db2772ad167bc7054f958411e7d6d8 M	salt
:040000 040000 7db42a0aa8e51421e4dba914a8db03c183328c3c 8b6002ad39f915a0bf96f3c7c0bd1041b81cfb24 M	tests
```

Looks like it was still you :P, I did it between 2015.5 -> develop

```
# bad: [909556ad268a327f5be8ebe06e785bdb7a02b047] Merge pull request #26282 from silenius/pw_group
# good: [a93b96c9ba1172cc288e5c94953b8b71bd3458f1] Merge pull request #26232 from garethgreenaway/2015_5_gpg_trust_key_fix
git bisect start 'develop' '2015.5'
# good: [b57da552ff521166650cfd366368455c37716a53] Merge pull request #26084 from twangboy/fix_25802
git bisect good b57da552ff521166650cfd366368455c37716a53
# good: [751081b1bcd6180604772ad07a55d02d3e6a3262] modules.boto_dynamodb: remove unused import
git bisect good 751081b1bcd6180604772ad07a55d02d3e6a3262
# good: [70d457484e685cef2f847b49c49ad770f2149083] Merge pull request #24686 from techhat/depreopenstack
git bisect good 70d457484e685cef2f847b49c49ad770f2149083
# good: [7a0bb9df1e1030e9ea0927806c0d3bf35b93641c] Merge pull request #25453 from isbm/isbm-libvirt
git bisect good 7a0bb9df1e1030e9ea0927806c0d3bf35b93641c
# good: [986935cc9f641b6f129ccc4f5537e5975be7228e] Merge pull request #25950 from s0undt3ch/features/raas-5-salt-ssh
git bisect good 986935cc9f641b6f129ccc4f5537e5975be7228e
# good: [860333755259b9afbb63795860f1402edc02c85b] Merge pull request #25984 from jmdcal/bugfix1
git bisect good 860333755259b9afbb63795860f1402edc02c85b
# bad: [1c661abc91c6a36ecd28338272007a5808edd27b] Skip failing event tests
git bisect bad 1c661abc91c6a36ecd28338272007a5808edd27b
# good: [5eda47aa8451ff8cc21d75cc9dd5e82421411701] Merge pull request #26145 from sjorge/imgadm-output-cleanup
git bisect good 5eda47aa8451ff8cc21d75cc9dd5e82421411701
# good: [82caf52497fa792f0eac20aceefbaafb4af925d2] Merge pull request #26201 from kev009/fbsd-disks-fix
git bisect good 82caf52497fa792f0eac20aceefbaafb4af925d2
# bad: [586411142979c7d499ed21021a4374dd4b2201d8] Minor spelling, style fixes
git bisect bad 586411142979c7d499ed21021a4374dd4b2201d8
# bad: [cfb698222713257f68a8697403bc67cea7b87fff] Adding some dictupdate merge tests
git bisect bad cfb698222713257f68a8697403bc67cea7b87fff
# good: [8edc0d82c61cada3648d0a82d8ec2df13b81eb0f] Merge pull request #26197 from TronPaul/gitfs-whitelist-base-fix
git bisect good 8edc0d82c61cada3648d0a82d8ec2df13b81eb0f
# good: [1c31722b494b70f50f20b399ab72ab3c49cc356f] Merge pull request #26211 from techhat/movespm
git bisect good 1c31722b494b70f50f20b399ab72ab3c49cc356f
# first bad commit: [cfb698222713257f68a8697403bc67cea7b87fff] Adding some dictupdate merge tests
```

**Comment by cachedout** [comment]
I'm confused. Before you said this was on 2015.3 but that's not in 2015.3?

**Comment by gtmanfred** [comment]
Somehow the error on 2015.5.3 disappeared... :/ i don't know why.  The same setups from last night that were failing, aren't failing now

The above is the result of a `git bisect start develop 2015.5`

**Comment by cachedout** [comment]
OK. I'll assume a stale .pyc or something.

So, with this. Do you have top files coming in from multiple salt envs or just a single top file?

**Comment by gtmanfred** [comment]
Just the single top file.  Here are my configs.

```
cat /etc/salt/master
interface: 172.16.0.1
fileserver_backend:
  - roots
timeout: 60
failhard: True
```
```
cat /etc/salt/cloud
update_cachedir: True
diff_cache_events: True
change_password: True
wait_for_ip_timeout: 6000
display_ssh_output: False
protocol: ipv4
script_args: -DLP git develop
minion:
  master: 172.16.0.1
  mine_functions:
    network.ip_addrs:
      - eth2
  tcp_keepalive: True
  tcp_keepalive_idle: 30
```
```
base:
  '*':
    - sshd
    - monitoring

  'salt*':
    - iptables.salt

  'board*.taplists.beer':
    - iptables
    - repos
    - redis
    - app
    - nginx
    - uwsgi
    - services
```

**Comment by cachedout** [comment]
OK, thanks. I'll start digging.

**Comment by gtmanfred** [comment]
Cool, if you want access to the environment, just let me know.

**Comment by ryan-lane** [comment]
Could this be related? https://github.com/saltstack/salt/issues/14723

**Comment by gtmanfred** [comment]
I just did one more check to make sure that the bisect worked correctly, and indeed the commit right before this one, does in fact work.

Thanks for looking @cachedout

**Comment by cachedout** [comment]
How very odd. Can I get access to this env? I'll hop on IRC.

**Comment by cachedout** [comment]
Fix is written. As @ryan-lane points out, we obviously need better testing here as well.

**Comment by cachedout** [comment]
Resolved by #26307
